third_party/xla_client/metrics_analysis.cc

```diff
@@ -0,0 +1,204 @@
+#include "tensorflow/compiler/xla/xla_client/metrics_analysis.h"
+
+#include "absl/types/variant.h"
+#include "tensorflow/compiler/xla/xla_client/computation_client.h"
+#include "tensorflow/compiler/xla/xla_client/metrics.h"
+#include "tensorflow/compiler/xla/xla_client/tf_logging.h"
+#include "tensorflow/compiler/xla/xla_client/types.h"
+
+namespace xla {
+namespace metrics {
+
+namespace {
+
+float DivToFloat(std::ldiv_t val, std::int64_t denominator) {
+  return val.quot + (float) val.rem / denominator;
+}
+
+class MetricFrequency : public Analyzer {
+ public:
+  MetricFrequency(std::string metric_name, float frequency_threshold,
+                  long warmup_steps = 0)
+      : metric_name_(metric_name),
+        frequency_threshold_(frequency_threshold),
+        warmup_steps_(warmup_steps) {}
+
+  Analysis Run() override {
+    CounterData* step = GetCounter("MarkStep");
+    MetricData* metric = GetMetric(metric_name_);
+    if (!step || !metric) {
+      return {Analysis::Symptom::kNormal};
+    }
+    size_t metric_count = metric->TotalSamples();
+    xla::int64 step_count = step->Value();
+    if (step_count <= warmup_steps_) {
+      return {Analysis::Symptom::kNormal};
+    }
+    auto res = std::div(metric_count, step_count);
+    if (DivToFloat(res, step_count) > frequency_threshold_) {
+      return {
+          Analysis::Symptom::kMetricTooFrequent,
+          absl::StrFormat("%s too frequent: %zu counts during %zu steps",
+                          metric_name_, metric_count, step_count),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::string metric_name_;
+  float frequency_threshold_;
+  long warmup_steps_;
+};
+
+class MetricTime : public Analyzer {
+ public:
+  MetricTime(std::string metric_name, long threshdold_nsec)
+      : metric_name_(metric_name), threshold_nsec_(threshdold_nsec) {}
+
+  Analysis Run() override {
+    double max_metric_time = 0;
+    MetricData* metric = GetMetric(metric_name_);
+    if (!metric) {
+      return {Analysis::Symptom::kNormal};
+    }
+    // No need for accumulator and we want all recent samples.
+    for (const Sample& sample : metric->Samples(nullptr, nullptr)) {
+      max_metric_time = std::max(sample.value, max_metric_time);
```

Comment:
maybe use [std::max_element](https://www.cplusplus.com/reference/algorithm/max_element/) here?

Reply:
As discussed, this is comparing `Sample` structs so code would be less readable so decided to stick to the above. 
```
    auto samples = metric->Samples(nullptr, nullptr);
    double max_metric_time =
        samples.empty()
            ? 0
            : (*std::max_element(samples.begin(), samples.end(),
                                 [](const Sample& first, const Sample& second) {
                                   return first.value < second.value;
                                 }))
                  .value;
```